### PR TITLE
fix shuffle_batch zero_division bug

### LINKED ISSUE #1853: shuffle_batch gives ZeroDivisionError when computing capacity stat
```
import tensorflow as tf
raw = tf.ones((2))
tf.train.shuffle_batch([raw], batch_size=2, capacity=4, min_after_dequeue=4, seed=5)

```

this fails with

```
ZeroDivisionError                         Traceback (most recent call last)
<ipython-input-11-86e4af7ca1ee> in <module>()
      1 import tensorflow as tf
      2 raw = tf.ones((2))
----> 3 tf.train.shuffle_batch([raw], batch_size=2, capacity=4, min_after_dequeue=4, seed=5)

/Users/yaroslav/anaconda/envs/tim-jan17/lib/python3.5/site-packages/tensorflow/python/training/input.py in shuffle_batch(tensors, batch_size, capacity, min_after_dequeue, num_threads, seed, enqueue_many, shapes, allow_smaller_final_batch, shared_name, name)
   1220       allow_smaller_final_batch=allow_smaller_final_batch,
   1221       shared_name=shared_name,
-> 1222       name=name)
   1223 
   1224 

/Users/yaroslav/anaconda/envs/tim-jan17/lib/python3.5/site-packages/tensorflow/python/training/input.py in _shuffle_batch(tensors, batch_size, capacity, min_after_dequeue, keep_input, num_threads, seed, enqueue_many, shapes, allow_smaller_final_batch, shared_name, name)
    779     full = (math_ops.cast(math_ops.maximum(0, queue.size() - min_after_dequeue),
    780                           dtypes.float32) *
--> 781             (1. / (capacity - min_after_dequeue)))
    782     # Note that name contains a '/' at the end so we intentionally do not place
    783     # a '/' after %s below.

```

Unlike TF tensors, Python doesn't handle float division by zero. One solution is to wrap "capacity" and "min_after_dequeue" into TF tensors so that you get "inf" as a result instead of RuntimeError

### PR DESCRIPTION
**Pull request by suiyuan2009**
fix #1853

### CODE REVIEW

**Comment by suiyuan2009** [comment]
@tensorflow-jenkins test this please

File: tensorflow/python/training/input.py [hunk 1]
```diff
@@ -723,7 +723,9 @@ def shuffle_batch(tensor_list, batch_size, capacity, min_after_dequeue,
     _enqueue(queue, tensor_list, num_threads, enqueue_many)
     full = (math_ops.cast(math_ops.maximum(0, queue.size() - min_after_dequeue),
                           dtypes.float32) *
-            (1. / (capacity - min_after_dequeue)))
+            math_ops.truediv(ops.convert_to_tensor(1,dtype=dtypes.float32) ,
+                ops.convert_to_tensor(capacity -
+                    min_after_dequeue,dtype=dtypes.float32)))
```
Comment (girving) [inline tensorflow/python/training/input.py]: Simplify this to

    tf.constant(1.) / (capacity - min_after_dequeue)

**Comment by girving** [comment]
Only admins are allowed to run the tests.  I'll fire them off once you make the suggested simplification.

**Comment by girving** [comment]
(admins -> organization members)

**Comment by suiyuan2009** [comment]
@girving add new commit, note that if `min_after_dequeue` equals `capacity`,  get
`tensorflow.python.framework.errors.InvalidArgumentError: min_after_dequeue 4 must be < capacity 4` instead of `inf`

File: tensorflow/python/training/input.py [hunk 2]
```diff
@@ -722,10 +722,8 @@ def shuffle_batch(tensor_list, batch_size, capacity, min_after_dequeue,
         dtypes=types, shapes=shapes, shared_name=shared_name)
     _enqueue(queue, tensor_list, num_threads, enqueue_many)
     full = (math_ops.cast(math_ops.maximum(0, queue.size() - min_after_dequeue),
-                          dtypes.float32) *
-            math_ops.truediv(ops.convert_to_tensor(1,dtype=dtypes.float32) ,
-                ops.convert_to_tensor(capacity -
-                    min_after_dequeue,dtype=dtypes.float32)))
+        dtypes.float32) *
+        (constant_op.constant(1.) / (capacity - min_after_dequeue)))
```
Comment (girving) [inline tensorflow/python/training/input.py]: Apologies: I should have noticed this before.  The code is `x * 1 / y`, but should be simplified to `x / y`.  Also, please fix the formatting to

    full = (math_ops.cast(math_ops.maximum(0, queue.size() - min_after_dequeue),
                          dtypes.float32) / (capacity - min_after_dequeue))

**Comment by girving** [comment]
Hmm, sorry I wasn't reading this carefully enough.  I think we should just throw a `ValueError` with a reasonable message if those two numbers are equal, since it is a useless queue.  @yaroslavvb: Since you filed the original bug, does that sound right to you?

**Comment by vrv** [comment]
ping for @girving and @yaroslavvb

**Comment by yaroslavvb** [comment]
@girving Correct, it's a useless queue. With my "convert_to_tensor" suggestion it will fail with "tensorflow/core/framework/op_segment.cc:53] Create kernel failed: Invalid argument: min_after_dequeue 4 must be < capacity 4" which is more informative than original division by zero in Summary construction. Throwing explicit ArgumentError sounds good to me too

**Comment by vrv** [comment]
Closing due to inactivity, comment and we'll reopen and retest, etc.

**Pull request closed without merging**